src/ramdb.cpp

```diff
@@ -0,0 +1,390 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2019 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/ramdb.hpp>
+
+#include <miopen/errors.hpp>
+#include <miopen/lock_file.hpp>
+#include <miopen/logger.hpp>
+#include <miopen/md5.hpp>
+
+#include <boost/filesystem/operations.hpp>
+#include <boost/filesystem.hpp>
+
+#include <chrono>
+#include <ctime>
+#include <fstream>
+#include <limits>
+#include <map>
+#include <mutex>
+#include <sstream>
+
+namespace miopen {
+
+std::string RamDb::GetTimeFilePath(const std::string& path) { return path + ".time"; }
+
+static ramdb_clock::time_point GetDbModificationTime(const std::string& path)
+{
+    const auto time_file_path = RamDb::GetTimeFilePath(path);
+    auto file                 = std::ifstream{time_file_path};
+
+    if(!file)
+        // Zero time from epoch, considering file ancient if time can't be retrieved.
+        return {};
+
+    ramdb_clock::rep time;
+    file >> time;
+    return ramdb_clock::time_point{ramdb_clock::duration{time}};
+}
+
+static void UpdateDbModificationTime(const std::string& path)
+{
+    MIOPEN_LOG_I2("Updating db modification time for " << path);
+
+    const auto time           = ramdb_clock::now().time_since_epoch();
+    const auto time_file_path = RamDb::GetTimeFilePath(path);
+    auto file                 = std::ofstream{time_file_path};
+
+    if(!file)
+    {
+        MIOPEN_LOG_E("Cannot update database modification time: " + time_file_path);
+        return;
+    }
+
+    file << time.count();
+}
+
+#define MIOPEN_VALIDATE_LOCK(lock)                       \
```

Comment:
It is defined in db.cpp, and I would prefer not to move it to db.hpp, which is indirectly included in almost the whole library due to the fact we use dbs as templates everywhere instead of just inheriting with an interface class or at least type erasure.